crates/sui-transactional-test-runner/src/programmable_transaction_test_parser/token.rs

```diff
@@ -0,0 +1,165 @@
+    }
+
+    fn next_token(s: &str) -> anyhow::Result<Option<(Self, usize)>> {
+        // parses a string where start matches end.
+        // performs simple matching for start/end pairs
+        fn parse_sub_token_string(
+            s: &str,
+            start: char,
+            end: char,
+        ) -> anyhow::Result<Option<usize>> {
+            // the length of the string until the matching end
+            let mut len = 0;
+            // the count of number of active start/end pairs
+            let mut count = 0i32;
+            let mut chars = s.chars().peekable();
+            loop {
+                let Some(c) = chars.next() else {
+                    bail!("Unexpected end of string after '('")
+                };
+                len += 1;
+                if c == start {
+                    // a new start char
+                    count += 1
+                } else if c == end {
+                    // an end char
+                    count -= 1;
+                    if count == 0 {
+                        // end found
+                        break;
+                    }
+                }
+            }
+            Ok(Some(len))
+        }
+
+        // type arguments get delegated to a different parser
+        if s.starts_with('<') {
+            return Ok(parse_sub_token_string(s, '<', '>')?.map(|len| (Self::TypeArgString, len)));
+        }
+        // comments
+        if s.starts_with("//") {
+            let end = s.find('\n').map(|end| end + 1).unwrap_or(s.len());
+            return Ok(Some((Self::Comment, end)));
+        }
+        if s.starts_with("/*") {
+            let Some(end) = s.find("*/") else {
+                bail!("Unexpected of string after start of comment '/*'")
+            };
+            return Ok(Some((Self::Comment, end + 2)));
+        }
+
+        // other tokens
+        let mut chars = s.chars().peekable();
+        let c = match chars.next() {
+            None => return Ok(None),
+            Some(c) => c,
+        };
+        Ok(Some(match c {
+            '(' => (Self::LParen, 1),
+            ')' => (Self::RParen, 1),
+            '[' => (Self::LBracket, 1),
+            ']' => (Self::RBracket, 1),
+            ',' => (Self::Comma, 1),
+            ';' => (Self::Semi, 1),
+            ':' if matches!(chars.peek(), Some(':')) => (Self::ColonColon, 2),
+            ':' => (Self::Colon, 1),
+            c if c.is_ascii_whitespace() => {
+                // c + remaining
+                let len = 1 + chars.take_while(char::is_ascii_whitespace).count();
+                (Self::Whitespace, len)
+            }
+            c if c.is_ascii_digit() => {
+                // c + remaining
+                let len = 1 + chars
+                    .take_while(|c| char::is_ascii_digit(c) || *c == '_')
+                    .count();
+                (CommandToken::Number, len)
+            }
+            c if c.is_ascii_alphabetic() => {
```

Comment:
What about underscores? (We can have valid identifiers in Move that start with underscores, right?)

Reply:
I don't think module or function names can start with them (from the Move compiler perspective). But I can add it